compiler/rustc_query_system/src/query/plumbing.rs

```diff
@@ -515,7 +514,13 @@ where
     // Some things are never cached on disk.
     if query.cache_on_disk {
         let prof_timer = tcx.dep_context().profiler().incr_cache_loading();
-        let result = query.try_load_from_disk(tcx, prev_dep_node_index);
+
+        // The call to `with_query_deserialization` enforces that no new `DepNodes`
+        // are created during deserialization. See the docs of that method for more
+        // details.
+        let result = dep_graph
+            .with_query_deserialization(|| query.try_load_from_disk(tcx, prev_dep_node_index));
```

Comment:
I think we can even forbid queries inside the invocation code. Can this call be moved around the `NotYetStarted` branch of `try_execute_query`? It's not supposed to invoke any user-provided code directly, only call into the `DepGraph` to setup a `TaskDeps`.

Reply:
I tried this earlier, but there are some invocations of `def_span` within the query infrastructure itself. I think those could be removed in a follow-up PR, since this PR should catch almost all of the cases.